turbopack/crates/turbo-rcstr/src/serde.rs

```diff
@@ -0,0 +1,97 @@
+use std::cell::RefCell;
+
+use indexmap::IndexSet;
+use rustc_hash::FxBuildHasher;
+use scoped_tls::scoped_thread_local;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use crate::RcStr;
+
+scoped_thread_local!(
+    /// Map of strings to their interned ids.
+    ///
+    /// This is used to serialize strings to their interned ids.
+    static SER_MAP: RefCell<IndexSet<RcStr,FxBuildHasher>>
+);
+
+scoped_thread_local!(
+    /// Read-only map of strings to their interned ids
+    static DE_MAP: IndexSet<RcStr,FxBuildHasher>
+);
+
+pub fn set_ser_map<F, R>(f: F) -> (R, IndexSet<RcStr, FxBuildHasher>)
+where
+    F: FnOnce() -> R,
+{
+    let map = Default::default();
+
+    let r = { SER_MAP.set(&map, f) };
+
+    (r, map.into_inner())
+}
+
+pub fn set_de_map<F, R>(map: &IndexSet<RcStr, FxBuildHasher>, f: F) -> R
+where
+    F: FnOnce() -> R,
+{
+    DE_MAP.set(map, f)
+}
+
+/// Intern a string for serialization.
+///
+/// This function exists to move the logic for accessing the SER_MAP out of the hot path of
+/// `Serialize::serialize`.
+#[inline(never)]
+fn intern_for_serialize(str: &RcStr) -> Option<u32> {
+    if !SER_MAP.is_set() {
+        return None;
+    }
+
+    Some(SER_MAP.with(|ser| {
+        let mut borrow = ser.borrow_mut();
+        if let Some(id) = borrow.get_index_of(str) {
+            id as u32
+        } else {
+            let id = borrow.len();
+            borrow.insert(str.clone());
+            id as u32
+        }
+    }))
+}
+
+impl Serialize for RcStr {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        if self.len() >= 3 && (self.len() < 512 || self.ref_count() >= 2) {
```

Comment:
using `self.ref_count()` here will break the determinism of interning.
This lead to the case where a key is sometimes interned and sometimes not, resulting in different serialized bytes, which will break the lookup.